### DISCUSSION TITLE
How to import CSS before the page load to prevent layout shift ?

### DISCUSSION THREAD

**Discussion by thlindustries** [vercel/next.js#26813] [Help]
## Problem history 
- Hello, I have a next project that uses the default next config for css/sass, 
I'm using node-sass to build a minified css file and i'm importing it into _app.js.
- Okay, so far no problem... The problem comes when I decided to get a minimum performance score of 80 points at lighthouse. Working on the performance i've discovered that my desktop styles are being imported even when it is at the mobile version of the webapp.
- My project uses a "Redirect Component" to detect if the client is on a mobile or desktop device, when the variable isMobile = true the project use components and containers from mobile folder, otherwise it uses from the desktop folder.

## The problem
- Using the coverage tool from devtools i realized that almost 180KB of css weren't used so I started trying to fix that.
- I changed my node-sass script to build separeted files of css, one for the mobile containers and components and an other one to desktop component and containers ... it went well and 180kb went  down to 78kb but there is one f*cking problem that 
I wasn't successful in finding out:

## What i've tried
- When I import the css files into _app I needed to import the correct file based on the value of isMobile variable so I started to search for conditional imports, sincerely i've not found to much about it so I did a ternary into the <Head> on _app :
```js
{this.props.isMobile ? (
  <link rel="preload" href="src/styles/build/css/Global.css" />
) : (
  <link rel="preload" href="src/mobile/styles/build/css/GlobalMobile.css" />
)}
```

## My results
... it went well and do the conditional import job, the problem is that there is a 'fast flash' of the page without the styles, and it causes the "layout shift" and because of that  I lose points in the lighthouse.

## Aditional Info
- I have access to the isMobile variable into the server side function, so at the first time that my _app renders the info is already available... Into my getInitialProps from _app you will see how I get the isMobile value:
```js
static async getInitialProps(appContext) {
   const mobxStore = initializeStores().mobxStores;
   mobxStore.productStore.cartId = cookies(appContext.ctx)[KEYS.CART_ID];
   appContext.ctx.mobxStore = mobxStore;

   const userAgent = appContext.ctx.req.headers['user-agent'];
   const isMobile = /(iPad|iPhone|Android|Mobile)/i.test(userAgent) || false;

   const appProps = await App.getInitialProps(appContext);
   return {
   	...appProps,
	initialMobxState: mobxStore,
	isMobile,
   };
}
```

## See what is happening:
Here is a gif to see what is happening :    https://imgur.com/WkqNrYT

# The Question
- Is there any work arround or solution to prevent this layout shift with dynamic imports? or is there any other way to do this and reduce my unused CSS to get lighthouse points?

**Comment by thlindustries** [comment]
any information that is missing just tell me that I update the topic

thank you first of all <3

**Reply by bitroy** [reply]
You can refer to this blog post, [CSS for Web Vitals](https://web.dev/css-web-vitals/)

If you do dynamic import of the component, make sure you use the loading component to preserve the space to prevent layout shifts, [next/dynamic custom loading component](https://nextjs.org/docs/advanced-features/dynamic-import#with-custom-loading-component)

**Reply by thlindustries** [reply]
cool, how could i refer them to this blog post?
and i think that you miss undesrtood my problem, in my case it's not an import that should happens in a specific time (using next dynamic import), my problem is with the conditional import, it's an import that should happen in my top level application but there is a variable that will define if the import is from mobileStyles or desktopStyles.
If the variable  *isMobile* is true, so the _app should import the  "src/mobile/styles/build/css/GlobalMobile.css" otherwise it should import "src/styles/build/css/Global.css"
